be/src/exprs/agg/array_agg.h

```diff
@@ -16,78 +16,192 @@
 #include "column/type_traits.h"
+#include "exec/sorting/sorting.h"
 #include "exprs/agg/aggregate.h"
+#include "exprs/function_context.h"
 #include "runtime/mem_pool.h"
 #include "types/logical_type.h"
 
 namespace starrocks {
 
-template <LogicalType LT>
+// input columns result in intermediate result: struct{array[col0], array[col1], array[col2]... array[coln]}
+// return ordered array[col0']
 struct ArrayAggAggregateState {
-    using ColumnType = RunTimeColumnType<LT>;
-
-    void update(const ColumnType& column, size_t offset, size_t count) { data_column.append(column, offset, count); }
-
-    void append_null() { null_count++; }
-    void append_null(size_t count) { null_count += count; }
-
-    ColumnType data_column; // Aggregated elements for array_agg
-    size_t null_count = 0;
+    void update(const Column& column, size_t index, size_t offset, size_t count) {
+        if (UNLIKELY(index >= data_columns->size())) {
+            throw std::runtime_error(
+                    fmt::format("index {} is larger than column size {}.", index, data_columns->size()));
+        }
+        (*data_columns)[index]->append(column, offset, count);
+    }
+    void update_nulls(size_t index, size_t count) {
+        if (UNLIKELY(index >= data_columns->size())) {
+            throw std::runtime_error(
+                    fmt::format("index {} is larger than column size {}.", index, data_columns->size()));
+        }
+        DCHECK((*data_columns)[index]->is_nullable());
+        (*data_columns)[index]->append_nulls(count);
+    }
+    ~ArrayAggAggregateState() {
+        if (data_columns != nullptr) {
+            for (auto& col : *data_columns) {
+                col.reset();
+            }
+            data_columns->clear();
+            delete data_columns;
+            data_columns = nullptr;
+        }
+    }
+    // using pointer rather than vector to avoid variadic size
+    // array_agg(a order by b, c, d), the a,b,c,d are put into data_columns in order.
+    Columns* data_columns;
 };
 
-template <LogicalType LT>
 class ArrayAggAggregateFunction
-        : public AggregateFunctionBatchHelper<ArrayAggAggregateState<LT>, ArrayAggAggregateFunction<LT>> {
+        : public AggregateFunctionBatchHelper<ArrayAggAggregateState, ArrayAggAggregateFunction> {
 public:
-    using InputColumnType = RunTimeColumnType<LT>;
+    void create(FunctionContext* ctx, AggDataPtr __restrict ptr) const override {
+        auto num = ctx->get_num_args();
+        auto* state = new (ptr) ArrayAggAggregateState;
+        state->data_columns = new Columns;
+        for (auto i = 0; i < num; ++i) {
+            auto arg_type = ctx->get_arg_type(i);
+            // TODO: support nested type
+            if (arg_type->type == LogicalType::TYPE_STRUCT || arg_type->type == LogicalType::TYPE_ARRAY ||
+                arg_type->type == LogicalType::TYPE_MAP) {
+                throw std::runtime_error(fmt::format("array_agg can't support nest type now."));
```

Comment:
create function is not thread safe we shouldn't throw any exception.

Reply:
move this check to FE.